### ISSUE TITLE
Warn if generator is iterated over more than once

### ISSUE THREAD

**Issue by thomkeh** [python/mypy#8640]
Somewhat related to #4707

This code currently does not give any warnings:

```python
gen = (3 * x for x in range(10))  # revealed type: Generator[int, None, None]
for i in gen:
    print(i)

for i in gen:
    print(i)
```

and while this does run, it doesn't do what the programmer expects. Only the first iteration prints anything. The second iteration stops immediately because the generator is already exhausted.

This might be very difficult, but it would be nice if mypy could detect this.

**Comment by silviogutierrez** [comment]
This just happened to me yesterday. Would be great to catch it.

However, I think this falls in the realm of unique types, which is about as advanced as type systems go. So I'm not sure mypy will ever land this...[1]


[1] https://en.wikipedia.org/wiki/Uniqueness_type

**Comment by JukkaL** [comment]
Mypy won't be able to catch all issues like these, but at least simple cases like the same generator being iterated twice in succession in a single scope would be possible to catch.

We might want to do this for all iterables, not just generators.
